:unicorn: added color segmentation

tohahn opened the pull request:
Added working color segmentation to robosherlock.

bbferka commented on src/segmentation/include/rs/HueClusterComparator.h:
```diff
@@ -0,0 +1,206 @@
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2010-2012, Willow Garage, Inc.
```
Did you implement this yourself or is it really from PCL?

tohahn commented:
the base code is from pcl. i added the code necessary for hue comparison and removed the code for specifying which points should be compared (labels).

bbferka commented:
do you think you could make this work on existing hypotheses as well?..as an option in the xml, the input cloud would then be a hypotheses that matches some description.

tohahn commented:
I think you mean the following: The cloud that should be clustered should
in some cases not be the scene but a robosherlock cluster that we think is
some object. That would be possible, but I need to know what the
description of the hypothis is and how to get the rs cluster based on the
description. Or do you mean something different?

bbferka commented:
exactly. The description of that hypotheses could be in the query:

`rs::Query q = rs::create<rs::Query>(tcas);`
`cas.getFS("QUERY",` q);`

you can find an example in the RegionFilter. Query has a field called asJson. and you can check there for an ID for example, which needs to match the idx of the cluster....for now this would suffice.

tohahn commented:
will do.

2017-05-03 14:48 GMT+02:00 Ferenc Balint-Benczedi <notifications@github.com>
:

> exactly. The description of that hypotheses could be in the query:
>
> rs::Query q = rs::create<rs::Query>(tcas);
> cas.getFS("QUERY", q);`
>
> you can find an example in the RegionFilter. Query has a field called
> asJson. and you can check there for an ID for example, which needs to match
> the idx of the cluster....for now this would suffice.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RoboSherlock/robosherlock/pull/86#issuecomment-298900959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVn9HSy8jvDP-sOhpAS1PC0sZlZ174Zyks5r2HejgaJpZM4M0Oie>
> .
>

tohahn commented:
i have two problems:
1. how do i get the id from the asJson?
2. how do I get the ID from a rs::Cluster?

2017-05-03 16:11 GMT+02:00 Tobias Hahn <morowinger@gmail.com>:

> will do.
>
> 2017-05-03 14:48 GMT+02:00 Ferenc Balint-Benczedi <
> notifications@github.com>:
>
>> exactly. The description of that hypotheses could be in the query:
>>
>> rs::Query q = rs::create<rs::Query>(tcas);
>> cas.getFS("QUERY", q);`
>>
>> you can find an example in the RegionFilter. Query has a field called
>> asJson. and you can check there for an ID for example, which needs to match
>> the idx of the cluster....for now this would suffice.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/RoboSherlock/robosherlock/pull/86#issuecomment-298900959>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AVn9HSy8jvDP-sOhpAS1PC0sZlZ174Zyks5r2HejgaJpZM4M0Oie>
>> .
>>
>
>

bbferka commented:
asJson is a json string...you can use a parser to look for an id field. for rs::Cluster you can simply use the index in the vector for now

tohahn commented:
The color clusterer now clusters for hue and for value. Also, it is possible to discretize values beforehand to avoid spilling. The possibility to cluster only on a hypothesis (id of a preexisting cluster) was also added.

bbferka commented:
looks ok, but you should not put rapidjson in the project. We don't like introducing dependencies that do not have a debian package. With rapidjson this will be solved once we migrate 16.04. If  this is simply there for the checking of the query for ID, then I suggest replacing it with a simple comparison e.g. asJson == "clusterNR". 

The other option is setting it as an external dependency, looking for rapidjson and conditioning the compilation of the annotator on this:
```cmake
find_package(RapidJSON QUIET)

if(${RapidJSON_FOUND})
  message("Found RapidJson, building XYZ Annotator")
  include_directories(${RAPIDJSON_INCLUDE_DIRS})
endif()
```

tohahn commented:
the lookup in the json string is now done with standard string functions.

bbferka commented:
hmm...this is an output that I get:
```
PointCloudColorSegmentation.cpp(159)[processWithLock] Found 1904 clusters for 1228800 points.
````
I should not end up with a point cloud of that size. It seems that you are concatenating point clouds, and keeping old indices as a result, or something similar (see picture).

![screenshot from 2017-06-21 12 48 01](https://user-images.githubusercontent.com/3843220/27380588-f515e21c-567f-11e7-904c-b661251aa1b4.png)

tohahn commented:
the latest commit was formatted with the styling guidelines and fixes the bug with old point cloud information by clearing the clouds before every iteration of the annotator.
